src/execution_tree/primitives/for_operation.cpp

```diff
@@ -0,0 +1,142 @@
+{
+    ///////////////////////////////////////////////////////////////////////////
+    std::vector<match_pattern_type> const for_operation::match_data =
+    {
+        hpx::util::make_tuple(
+            "for", "for(_1, _2, _3, _4)", &create<for_operation>)
+    };
+
+    ///////////////////////////////////////////////////////////////////////////
+    for_operation::for_operation(
+            std::vector<primitive_argument_type>&& operands)
+      : operands_(std::move(operands))
+    {
+        if (operands_.size() != 4)
+        {
+            HPX_THROW_EXCEPTION(hpx::bad_parameter,
+                "phylanx::execution_tree::primitives::for_operation::"
+                    "for_operation",
+                "the for_operation primitive requires exactly four arguments");
+        }
+
+        if (!valid(operands_[0]) || !valid(operands_[1])
+            || !valid(operands_[2]) || !valid(operands_[3]))
+        {
+            HPX_THROW_EXCEPTION(hpx::bad_parameter,
+                "phylanx::execution_tree::primitives::for_operation::"
+                    "for_operation",
+                "the for_operation primitive requires that the arguments "
+                    "given by the operands array are valid");
+        }
+    }
+
+    namespace detail
+    {
+        struct iteration : std::enable_shared_from_this<iteration>
+        {
+            iteration(std::vector<primitive_argument_type> const& operands)
+              : operands_(operands)
+            {}
+
+          hpx::future<primitive_result_type> init()
+          {
+            auto this_ = this->shared_from_this();
+            return literal_operand(operands_[0]).then(
+                [this_](auto val)
+                {
+                  val.get(); //this future should already be ready and hence not block
+                  return this_->loop();
+                });
+          }
+
+          hpx::future<primitive_result_type> reinit()
+          {
+            auto this_ = this->shared_from_this();
+            return literal_operand(operands_[2]).then(
+                [this_](auto val)
+                {
+                  return val.get(); //this future should already be ready
+                  // and hence not block
+                });
+          }
+
+            hpx::future<primitive_result_type> body(
+                hpx::future<primitive_result_type>&& cond)
+            {
+                if (extract_boolean_value(cond.get()))
+                {
+                    // evaluate body of for statement
+                    auto this_ = this->shared_from_this();
+                    return literal_operand(operands_[3]).then(
+                        [this_](
+                            hpx::future<primitive_result_type> && result
+                        ) mutable
+                        {
+                            this_->result_ = result.get();
+                            //do the reinit statement
+                            auto hold = this_->reinit().get();
```

Comment:
Wouldn't it be better to evaluate the reinit asynchronously as well?

Reply:
@hkaiser  Yes, that needs fixing. I will fix it.